### PR TITLE
Pfs 117 upload page

### PR DESCRIPTION
**Pull request by Nicholassully**

### CODE REVIEW

File: internal/api/upload.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/opg-sirius-finance-admin/internal/model"
+	"net/http"
+	"os"
+)
+
+func (c *ApiClient) Upload(ctx Context, reportUploadType string, uploadDate string, email string, file *os.File) error {
```
Comment (hawx) [inline internal/api/upload.go:11]: as I commented on `Download`, I think you could extract things to a `model.NewUpload` function, and use `switch` at the end

File: internal/api/upload.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		ReportUploadType: reportUploadType,
+		UploadDate:       uploadDateTransformed,
+		Email:            email,
+		File:             file,
+	})
+	if err != nil {
+		return err
+	}
+
+	if reportUploadType == "DebtChase" {
```
Comment (hawx) [inline internal/api/upload.go:31]: also I'd be tempted to refactor this to a `switch` that sets `reqPath` or something as that is the only thing that changes, then construct the request (or extract as a function)

File: internal/api/upload.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	if uploadDate != "" {
+		uploadDateFormatted := model.NewDate(uploadDate)
+		uploadDateTransformed = &uploadDateFormatted
+	}
+
+	err := json.NewEncoder(&body).Encode(model.Upload{
+		ReportUploadType: reportUploadType,
+		UploadDate:       uploadDateTransformed,
+		Email:            email,
+		File:             file,
```
Comment (hawx) [inline internal/api/upload.go:25]: does this actually work with `*os.File`? I'm guessing really this is expecting some base64 version of the contents, and then you'd be able to change the `file` param above to `io.Reader` or some simpler type

File: internal/api/upload_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	if _, err := tempFile.Write(content); err != nil {
+		t.Fatal(err)
+	}
+	// Reset the file pointer to the beginning
+	if _, err := tempFile.Seek(0, io.SeekStart); err != nil {
+		t.Fatal(err)
+	}
+
+	// Define test cases
+	testCases := []struct {
```
Comment (hawx) [inline internal/api/upload_test.go:40]: could simplify this to `map[string]string`

File: internal/server/environment_vars.go [hunk 5]
```diff
@@ -19,6 +20,7 @@ func NewEnvironmentVars() EnvironmentVars {
 		SiriusURL:       getEnv("SIRIUS_URL", "http://host.docker.internal:8080"),
 		SiriusPublicURL: getEnv("SIRIUS_PUBLIC_URL", ""),
 		Prefix:          getEnv("PREFIX", ""),
+		BackendUrl:      getEnv("BACKEND_URL", ""),
```
Comment (hawx) [inline internal/server/environment_vars.go:23]: this should be `BackendURL` to match naming conventions

File: internal/api/api_client.go [hunk 6]
```diff
@@ -21,18 +22,22 @@ func (e ClientError) Error() string {
 	return string(e)
 }
 
-func NewApiClient(httpClient HTTPClient) *ApiClient {
+func NewApiClient(httpClient HTTPClient, siriusUrl string, backendUrl string) (*ApiClient, error) {
```
Comment (hawx) [inline internal/api/api_client.go:25]: rename these to `siriusURL` and `backendURL` to match convention

File: internal/server/upload.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+	// Handle file upload
+	file, headers, err := r.FormFile("fileUpload")
+	if err != nil {
+		return h.handleError(w, r, "No file uploaded", http.StatusBadRequest)
+	}
+	defer file.Close()
+
+	// Define the destination file path
+	dst, err := os.Create(headers.Filename)
```
Comment (hawx) [inline internal/server/upload.go:50]: do you need to create a real file here, or would you be able to get away with using the data?

File: internal/server/upload.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+	return nil
+}
+
+func reportHeadersByType(reportType string) []string {
+	var headers []string
+
+	switch reportType {
+	case model.ReportTypeUploadDeputySchedule.Key():
+		headers = []string{"Deputy number", "Deputy name", "Case number", "Client forename", "Client surname", "Do not invoice", "Total outstanding"}
```
Comment (hawx) [inline internal/server/upload.go:101]: could early return in each of these branches, instead of setting a var

File: internal/api/upload.go (hunk 1 above)
Comment (Nicholassully) [reply internal/api/upload.go]: done

File: internal/api/upload.go (hunk 2 above)
Comment (Nicholassully) [reply internal/api/upload.go]: done

File: internal/api/upload_test.go (hunk 4 above)
Comment (Nicholassully) [reply internal/api/upload_test.go]: done

File: internal/server/environment_vars.go (hunk 5 above)
Comment (Nicholassully) [reply internal/server/environment_vars.go]: done

File: internal/api/api_client.go (hunk 6 above)
Comment (Nicholassully) [reply internal/api/api_client.go]: done

File: internal/server/upload.go (hunk 8 above)
Comment (Nicholassully) [reply internal/server/upload.go]: done

**Review by jackgoodby — DISMISSED** [review]

**Review by jackgoodby — APPROVED** [review]

**Pull request merged**